File: backend/src/main/java/dmei22/flashy/service/CardService.java
package dmei22.flashy.service;

import dmei22.flashy.dto.card.CardDetailsDto;
import dmei22.flashy.dto.card.CardUpdateRequest;
import dmei22.flashy.dto.review.ReviewCreateRequest;
import dmei22.flashy.model.Card;
import dmei22.flashy.model.Level;
import dmei22.flashy.model.Review;
import dmei22.flashy.repository.CardRepository;
import dmei22.flashy.repository.ReviewRepository;
import dmei22.flashy.service.mapper.CardMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;

@Service
@Transactional
public class CardService {

    private final CardRepository cardRepository;
    private final ReviewRepository reviewRepository;

    public CardService(
            CardRepository cardRepository,
            ReviewRepository reviewRepository
    ) {
        this.cardRepository = cardRepository;
        this.reviewRepository = reviewRepository;
    }

    // CREATE
    public void createReview(Long cardId, ReviewCreateRequest request) {
        Card card = this.cardRepository.findById(cardId).get();
        Review review = new Review();

        review.setCard(card);
        review.setReviewDate(LocalDate.now());
        review.setSucces(request.isSuccess());
        card.getReviews().add(review);
        if (request.isSuccess()) {
            card.setLevel(card.getLevel().next());
            card.setDueDate(LocalDate.now().plusDays(card.getLevel().getDays()));
        } else {
            card.setLevel(Level.NEW);
            card.setDueDate(LocalDate.now());
        }

        this.reviewRepository.save(review);
    }

    // READ
    public CardDetailsDto getCard(Long cardId) {
        Card card = this.cardRepository.findById(cardId).get();
        return CardMapper.toCardDetailsDto(card);
    }

    // UPDATE
    public CardDetailsDto updateCard(
            Long cardId,
            CardUpdateRequest request
    ) {
        Card card = this.cardRepository.findById(cardId).get();
        card.setFront(request.getFront());
        card.setBack(request.getBack());

        return CardMapper.toCardDetailsDto(this.cardRepository.save(card));
    }

    // DELETE
}

File: frontend/src/app/service/card.service.ts
import {inject, Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {CardCreateRequest, CardDetails, CardOverview, CardUpdateRequest} from "../model/Card";
import {ReviewCreateRequest} from "../model/Review";

@Injectable({
  providedIn: 'root'
})
export class CardService {

  private http = inject(HttpClient);
  private requestMapping = "http://localhost:8080/card"

  constructor() { }

  // CREATE
  public createReview(cardId: number, request: ReviewCreateRequest): Observable<void> {
    return this.http.post<void>(`${this.requestMapping}/review`, request);
  }

  // READ
  public getCard(cardId: number): Observable<CardDetails> {
    return this.http.get<CardDetails>(`${this.requestMapping}/${cardId}`);
  }

  // UPDATE
  public update(cardId: number, request: CardUpdateRequest): Observable<CardDetails> {
    return this.http.put<CardDetails>(`${this.requestMapping}/${cardId}`, request);
  }

  // DELETE

}
